lib/src/klio/utils/_thread_limiter.py

```diff
@@ -0,0 +1,157 @@
+
+    .. code-block:: python
+
+        # No arg to use the default limit (CPU count of host/worker):
+        thread_limiter = ThreadLimiter()
+
+        # Then use it as a context manager:
+        with thread_limiter:
+            ...
+
+        # Or explicitly call acquire and release:
+        thread_limiter.acquire()
+        ...
+        thread_limiter.release()
+
+        # Explicitly use the default limit:
+        thread_limiter = ThreadLimiter(max_thread_count=ThreadLimit.DEFAULT)
+
+        # Set your own limit:
+        thread_limiter = ThreadLimiter(max_thread_count=2)
+
+        # Dynamically set the limit using a function
+        limit_func = lambda: multiprocessing.cpu_count() * 4
+        thread_limiter = ThreadLimiter(max_thread_count=limit_func)
+
+        # Turn off thread limiting
+        thread_limiter = ThreadLimiter(max_thread_count=ThreadLimit.NONE)
+
+    Args:
+        max_thread_count (int, callable, ThreadLimit): number of threads
+            to make available to the limiter, or a :func:`callable` that
+            returns an ``int``. Set to :attr:`ThreadLimit.NONE` for no
+            thread limits.
+        name (str): Name of particular limiter. Defaults to object ID via
+            ``id(self)``.
+    """
+
+    _PREFIX = "KlioThreadLimiter"
+
+    def __init__(self, max_thread_count=ThreadLimit.DEFAULT, name=None):
+        self.__repr_name = f"name={name}" if name else f"id={id(self)}"
+        self.logger = logging.getLogger("klio.concurrency")
+
+        if max_thread_count is ThreadLimit.DEFAULT:
+            max_thread_count = multiprocessing.cpu_count
+
+        if max_thread_count is ThreadLimit.NONE:
+            self._dummy = True
+            self._semaphore = _DummySemaphore()
+            self.logger.info(f"{self} Using unlimited semaphore")
+
+        else:
+            self._dummy = False
+            if callable(max_thread_count):
+                max_thread_count = max_thread_count()
+            self._semaphore = threading.BoundedSemaphore(max_thread_count)
+            self.logger.info(
+                f"{self} Initial semaphore value: {self._semaphore._value}"
+            )
+
+    def _log(self, message):
+        threads_available = False
+        suffix_msg = ""
+        if not self._dummy:
+            threads_available = self._semaphore._value
+            suffix_msg = f" (available threads: {threads_available})"
+        self.logger.info(f"{self} {message}{suffix_msg}")
+
+    def acquire(self):
+        """Acquire a semaphore (a thread).
+
+        Acquiring a semaphore will activate an available thread in Beam's
+        multi-threaded environment.
+
+        If no semaphores are available, the method will block until one
+        is released via :func:`ThreadLimiter.release`.
+        """
+        self._log("Blocked – waiting on semaphore for an available thread")
+        self._semaphore.acquire()
+
+    def release(self):
+        """Release a semaphore (a thread)."""
```

Comment:
If we're not explicitly calling out the fact that we're using a BoundedSemaphore under the hood, it could be good to say that when the thread_limiter is released too many times, a ValueError will be raised.